### ISSUE TITLE
improve twistd logging setup factoring to remove duplicate code and simplify extending it

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2574]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#2574|
|Type|task|
|Created|2007-04-16 00:24:02Z|

twistd and twistw have slightly different codepaths for logging setup since they each implement it themselves.  The factoring also isn't friendly towards efforts to allow custom log observers to be supplied.

The code should be refactored to remove the duplication and make it simpler to override the default behavior.

<details><summary>Searchable metadata</summary>

```
trac-id__2574 2574
type__task task
reporter__exarkun exarkun
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1176683042000000 1176683042000000
changetime__1177256675000000 1177256675000000
version__None None
owner__ 
cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun removed owner|
|-|-|

Code for this in refactor-twistd-logging-2574.  This will conflict with the [#11359](https://github.com/twisted/twisted/issues/11359) branch, but it will ultimately simplify the code necessary to resolve that ticket.

This is a short, self-contained change and should be simple to review, although since this is twistd code there may well be obscure problems lurking in the shadows.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve set owner to @exarkun|
|-|-|

That's great, a good move towards where we want to go. I have only one remark: why use getLogObserver function in _twistd_unix and _twistw instead of defining it directly in the ApplicatinoRunner class ?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set owner to @therve|
|-|-|

Good question.  I got rid of the free functions and put the code directly into the methods on the subclasses.  I also added a couple simple unit tests.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

Hrm.

I wonder if it would be a good idea to leave the startLogging API in place, deprecated, for a while.  Probably.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

Okay I re-added startLogging to _twistd_unix.py and _twistw.py with pending deprecation warnings.  I think that was actually the reason I left the free-functions in before, I just forgot to do it. ;)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set owner to @exarkun|
|-|-|

Hmm some unit tests are failing in the branch though, not sure why

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun removed owner|
|-|-|

Okay fixed them, code is still in refactor-twistd-logging-2574

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve set owner to @exarkun|
|-|-|

OK, I'm not really happy with _postApplication/postApplication. Does it shock you to have preApplication/loadApp/startLogging/postApplication in the run method of the ApplicationRunner?

Otherwise that looks good.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set status to `closed`|
|-|-|

(In [20034]) Merge refactor-twistd-logging-2574

Author: exarkun
Reviewer: therve
Fixes [#2574](https://github.com/twisted/twisted/issues/2574)

Move the logging system initialization calls out of the platform-specific
ApplicationRunner subclasses and into the base class so that it can be
customized more easily.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
